src/future/join/tuple.rs

```diff
@@ -21,8 +92,10 @@ macro_rules! impl_merge_tuple {
         #[must_use = "futures do nothing unless you `.await` or poll them"]
         #[allow(non_snake_case)]
         pub struct $StructName<$($F: Future),*> {
-            done: bool,
-            $(#[pin] $F: MaybeDone<$F>,)*
+            len: u32,
+            $(#[pin] $F: $F,)*
```

Comment:
Would it be possible to instead do:

```suggestion
            tuple: ($(#[pin] $F: $F,)*),
```

That way we can move the tuple fields into here basically in-place.

Reply:
I'm currently trying to figure out how to do this!

`tuple: ($(#[pin] $F: $F,)*)` doesn't work because `pin_project` doesn't support pinning this way, like `(#[pin] T, #[pin] S)`

Reply:
@yoshuawuyts, I got something!

It's on a second branch, [here](https://github.com/yoshuawuyts/futures-concurrency/compare/main...matheus-consoli:futures-concurrency:remove-maybedone-from-join-tuple-take2).

I took some inspiration from #84 to use the pinned futures inside a tuple(struct), and it kinda works.

We have some performance gain, but it's kinda negligible :disappointed: (especially because of the macro-heavy impl orientation)

```text
>> critcmp main patch -f tuple::join
group             main                                   patch
-----             ----                                   -----
tuple::join 10    1.03    239.3±1.06ns        ? ?/sec    1.00    231.3±0.79ns        ? ?/sec
```

Reply:
Oh yeah, I like the intermediate struct approach!

The reason why there's a small perf cost might be because you're using `ptr::read` there which [ends up doing an extra copy]((https://doc.rust-lang.org/src/core/ptr/mod.rs.html#1117))? Instead I think if it's at all possible it might be worth attempting to perform a direct transmute? This might require adding a `#[repr(transparent)]` to the `Futures` struct so that the layout is guaranteed to be the same as the tuples it stores?

Reply:
oh, I don't think `#[repr(transparent)] Futures` is possible, `transparent` only allows one field to have a non-zero size, the `Futures` have many.

Reply:
I'm trying to `transmute`, but maybe I have found a limitation (?)

I can transmute from `(MaybeUninit<ConcreteTypeA>, MaybeUninit<ConcreteTypeB>)` to `(ConcreteTypeA, ConcreteTypeB)`, but I can't to the same from `(MaybeUninit<GenericTypeA>, MaybeUninit<GenericTypeB>)`, [play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e948fe4a441734c81c5ce7b35290a23b)

I'll try again after lunch

Reply:
> oh, I don't think `#[repr(transparent)] Futures` is possible, `transparent` only allows one field to have a non-zero size, the `Futures` have many.

oof, TIL :')

Reply:
oook, it seems to me that we cannot transmute `(MaybeUninit<T>, MaybeUninit<U>)` to `(T, U)`.

I'm not sure why, but I guess it's the same underlying problem from rust-lang/rust#61956, so casting the tuples appears to be the best option so far

Reply:
Transmuting is not what you want here I think ?

```rust
let t = (MaybeUninit::<usize>::new(42), MaybeUninit::<u8>::new(12)); // source, can come from anywhere

/* ... */

// Where you wanted to transmute:
let (a, b) = t;
(a.assume_init(), b.assume_init())
```

This can be macro-ified using `$F` as the variable name: it won't even shadow the type names because that's two different namespaces (`let usize: usize = 4_usize;` is valid)

Reply:
Assuming the `t` is owned, this should not do any copy anywhere

Reply:
ohh, thank you for pointing it out! I'll try that